test/fastify-session/passport.test.ts

```diff
@@ -0,0 +1,459 @@
+  server.get(
+    '/',
+    { preValidation: fastifyPassport.authenticate('test', { authInfo: true }) },
+    async () => 'hello world!'
+  )
+  server.post('/login', fastifyPassport.authenticate('test', { successRedirect: '/', authInfo: true }))
+
+  const login = await server.inject({
+    method: 'POST',
+    payload: { login: 'test', password: 'test' },
+    url: '/login',
+  })
+  expect(login.statusCode).toEqual(302)
+  expect(login.headers.location).toEqual('/')
+
+  const response = await server.inject({
+    url: '/',
+    headers: {
+      cookie: login.headers['set-cookie'],
+    },
+    method: 'GET',
+  })
+
+  expect(response.body).toEqual('hello world!')
+  expect(response.statusCode).toEqual(200)
+})
+
+test(`should not log the user in when passed a callback`, async () => {
+  const { server, fastifyPassport } = getConfiguredTestServer()
+  server.get(
+    '/',
+    { preValidation: fastifyPassport.authenticate('test', { authInfo: true }) },
+    async () => 'hello world!'
+  )
+  server.post(
+    '/login',
+    fastifyPassport.authenticate('test', async (request, reply, err, user) => {
+      return (user as any).name
+    })
+  )
+
+  const login = await server.inject({
+    method: 'POST',
+    payload: { login: 'test', password: 'test' },
+    url: '/login',
+  })
+  expect(login.statusCode).toEqual(200)
+  expect(login.body).toEqual('test')
+
+  const headers: Record<string, any> = {}
+  if (login.headers['set-cookie']) {
+    headers['cookie'] = login.headers['set-cookie']
+  }
+
+  const response = await server.inject({
+    url: '/',
+    headers,
+    method: 'GET',
+  })
+
+  expect(response.statusCode).toEqual(401)
+})
+
+test(`should allow registering strategies after creating routes referring to those strategies by name`, async () => {
+  const { server, fastifyPassport } = getRegisteredTestServer()
+
+  server.get(
+    '/',
+    { preValidation: fastifyPassport.authenticate('test', { authInfo: false }) },
+    async (request, reply) => {
+      void reply.send(request.session['messages'])
+    }
+  )
+
+  server.post(
+    '/login',
+    {
+      preValidation: fastifyPassport.authenticate('test', {
+        successRedirect: '/',
+        successMessage: 'welcome',
+        authInfo: false,
+      }),
+    },
+    () => {}
+  )
+
+  // register the test strategy late (after the above .authenticate calls)
+  fastifyPassport.use(new TestStrategy('test'))
+
+  const loginResponse = await server.inject({
+    method: 'POST',
+    url: '/login',
+    payload: { login: 'test', password: 'test' },
+  })
+
+  expect(loginResponse.statusCode).toEqual(302)
+  expect(loginResponse.headers.location).toEqual('/')
+
+  const homeResponse = await server.inject({
+    url: '/',
+    headers: {
+      cookie: loginResponse.headers['set-cookie'],
+    },
+    method: 'GET',
+  })
+
+  expect(homeResponse.body).toEqual('["welcome"]')
+  expect(homeResponse.statusCode).toEqual(200)
+})
```

Comment:
Can we avoid duplicating all these tests, and instead wrap them up in a function that gets invoked twice in the appropriate different contexts? Its a bunch more maintenance burden to have to edit 2x as many tests, and other contributors might not know they need to add tests to both places.